private/buf/bufgen/config.go

```diff
@@ -16,51 +16,89 @@ package bufgen
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"os"
 	"path/filepath"
 
-	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
-	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginref"
-	"github.com/bufbuild/buf/private/bufpkg/bufremoteplugin"
 	"github.com/bufbuild/buf/private/pkg/encoding"
-	"github.com/bufbuild/buf/private/pkg/normalpath"
 	"github.com/bufbuild/buf/private/pkg/storage"
 	"go.uber.org/zap"
-	"google.golang.org/protobuf/types/descriptorpb"
 )
 
-func readConfig(
+func newReadConfigOptions() *readConfigOptions {
+	return &readConfigOptions{}
+}
+
+type readConfigOptions struct {
+	override string
+}
+
+func readConfigVersion(
 	ctx context.Context,
 	logger *zap.Logger,
-	provider Provider,
+	provider ConfigDataProvider,
 	readBucket storage.ReadBucket,
 	options ...ReadConfigOption,
-) (*Config, error) {
+) (string, error) {
+	version, err := ReadFromConfig(
+		ctx,
+		logger,
+		provider,
+		readBucket,
+		getConfigVersion,
+		options...,
+	)
+	if err != nil || version == nil {
+		return "", err
+	}
+	return *version, nil
+}
+
+func readFromConfig[V any](
+	ctx context.Context,
+	logger *zap.Logger,
+	provider ConfigDataProvider,
+	readBucket storage.ReadBucket,
+	configGetter ConfigGetter[V],
+	options ...ReadConfigOption,
+) (*V, error) {
 	readConfigOptions := newReadConfigOptions()
 	for _, option := range options {
 		option(readConfigOptions)
 	}
 	if override := readConfigOptions.override; override != "" {
 		switch filepath.Ext(override) {
 		case ".json":
-			return getConfigJSONFile(logger, override)
+			return getConfigJSONFile(logger, override, configGetter)
 		case ".yaml", ".yml":
-			return getConfigYAMLFile(logger, override)
+			return getConfigYAMLFile(logger, override, configGetter)
 		default:
-			return getConfigJSONOrYAMLData(logger, override)
+			return getConfigJSONOrYAMLData(logger, override, configGetter)
 		}
 	}
-	return provider.GetConfig(ctx, readBucket)
+	data, id, err := provider.GetConfigData(ctx, readBucket)
+	if err != nil {
+		return nil, err
+	}
+	return configGetter(
+		logger,
+		encoding.UnmarshalYAMLNonStrict,
+		encoding.UnmarshalYAMLStrict,
+		data,
+		id,
```

Comment:
The `id` here is the reason why interface `ConfigDataProvider`'s `GetConfigData` returns an id.